squidpy/pl/_spatial_utils.py

```diff
@@ -0,0 +1,897 @@
+        color_source_vector = adata.raw.obs_vector(value_to_plot)
+    else:
+        color_source_vector = adata.obs_vector(value_to_plot, layer=layer)
+    if groups is not None and is_categorical_dtype(color_source_vector):
+        color_source_vector = color_source_vector.replace(color_source_vector.categories.difference(groups), np.nan)
+
+    to_hex = partial(colors.to_hex, keep_alpha=True)
+    if value_to_plot is None:
+        return None, np.full(to_hex(na_color), adata.n_obs), False
+    if not is_categorical_dtype(color_source_vector):
+        return None, color_source_vector, False
+    else:
+        clusters = color_source_vector.categories
+        color_map = _get_palette(adata, cluster_key=value_to_plot, categories=clusters, palette=palette)
+        color_vector = color_source_vector.rename_categories(color_map)
+        # Set color to 'missing color' for all missing values
+        if color_vector.isna().any():
+            color_vector = color_vector.add_categories([to_hex(na_color)])
+            color_vector = color_vector.fillna(to_hex(na_color))
+        return color_source_vector, color_vector, True
+
+
+def _shaped_scatter(
+    x: NDArrayA,
+    y: NDArrayA,
+    s: float,
+    c: NDArrayA,
+    shape: _AvailShapes | None = ScatterShape.CIRCLE.s,  # type: ignore[assignment]
+    norm: _Normalize = None,
+    **kwargs: Any,
+) -> PatchCollection:
+    """
+    Get shapes for scatterplot.
+
+    Adapted from here: https://gist.github.com/syrte/592a062c562cd2a98a83 .
+    This code is under [The BSD 3-Clause License](http://opensource.org/licenses/BSD-3-Clause)
+    """
+    shape = ScatterShape(shape)  # type: ignore[assignment]
+    if TYPE_CHECKING:
+        assert isinstance(shape, ScatterShape)
+    shape = shape.s
+
+    if shape == ScatterShape.CIRCLE.s:
+        zipped = np.broadcast(x, y, s)
+        patches = [Circle((x_, y_), s_) for x_, y_, s_ in zipped]
+    else:
+        n = 4 if shape == ScatterShape.SQUARE.s else 6
+        r: float = s / (2 * np.sin(np.pi / n))
+        polys = np.stack([_make_poly(x, y, r, n, i) for i in range(n)], 1).swapaxes(0, 2)
+        patches = [Polygon(p, False) for p in polys]
+    collection = PatchCollection(patches, **kwargs)
+
+    if isinstance(c, np.ndarray) and np.issubdtype(c.dtype, np.number):
+        collection.set_array(np.ma.masked_invalid(c))
+        collection.set_norm(norm)
+    else:
+        collection.set_facecolor(c)
+
+    return collection
+
+
+def _make_poly(x: NDArrayA, y: NDArrayA, r: float, n: int, i: int) -> Tuple[NDArrayA, NDArrayA]:
+    x_i = x + r * np.sin((np.pi / n) * (1 + 2 * i))
+    y_i = y + r * np.cos((np.pi / n) * (1 + 2 * i))
+    return x_i, y_i
+
+
+def _plot_edges(
+    adata: AnnData,
+    coords: NDArrayA,
+    ax: Axes,
+    edges_width: float = 0.1,
+    edges_color: str | Sequence[float] | Sequence[str] = "grey",
+    connectivity_key: str = Key.obsp.spatial_conn(),
+    edges_kwargs: Mapping[str, Any] = MappingProxyType({}),
```

Comment:
Any reason to use `edges_kwargs` instead of `**kwargs`?

Reply:
no, good point